prow/cmd/sinker/main_test.go

```diff
@@ -1014,3 +1017,108 @@ func (c *clientWrapper) Get(ctx context.Context, key ctrlruntimeclient.ObjectKey
 	}
 	return c.Client.Get(ctx, key, obj)
 }
+
+func TestGetConfigMapSize(t *testing.T) {
+	toplevel, err := os.MkdirTemp("", "job-config")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(toplevel)
+
+	timestamp_dir := filepath.Join(toplevel, "..2024_01_01")
+	err = os.Mkdir(timestamp_dir, 0750)
+	if err != nil && !os.IsExist(err) {
+		t.Fatal(err)
+	}
+
+	// Create files inside timestamp_dir, just like how K8s does it.
+	file1 := filepath.Join(timestamp_dir, "key1")
+	if err := os.WriteFile(file1, []byte("val1"), 0666); err != nil {
+		t.Fatal(err)
+	}
+
+	file2 := filepath.Join(timestamp_dir, "key2")
+	if err := os.WriteFile(file2, []byte("val2"), 0666); err != nil {
+		t.Fatal(err)
+	}
+
+	file3 := filepath.Join(timestamp_dir, "key3")
+	if err := os.WriteFile(file3, []byte("val3"), 0666); err != nil {
+		t.Fatal(err)
+	}
+
+	// Symlink ..data to point to timestamp_dir.
+	data_dir := filepath.Join(toplevel, "..data")
+	os.Symlink(timestamp_dir, data_dir)
+
+	// Create symlinks at the toplevel that point to files in ..data
+	// (again, like how K8s does it).
+	os.Symlink(filepath.Join(data_dir, "key1"), filepath.Join(toplevel, "key1"))
+	os.Symlink(filepath.Join(data_dir, "key2"), filepath.Join(toplevel, "key2"))
+	os.Symlink(filepath.Join(data_dir, "key3"), filepath.Join(toplevel, "key3"))
+
+	gotBytes, err := getConfigMapSize(toplevel)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Expect 12 bytes, because the 3 files each have 4 bytes of content.
+	if gotBytes != 12 {
+		t.Errorf("expected 12 bytes but got %v", gotBytes)
+	}
+}
+
+func TestGetConfigMapDirs(t *testing.T) {
+	toplevel, err := os.MkdirTemp("", "job-config")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(toplevel)
+
+	subdir1 := filepath.Join(toplevel, "part-1")
+	err = os.Mkdir(subdir1, 0750)
+	if err != nil && !os.IsExist(err) {
+		t.Fatal(err)
+	}
+
+	subdir2 := filepath.Join(toplevel, "part-2")
+	err = os.Mkdir(subdir2, 0750)
+	if err != nil && !os.IsExist(err) {
+		t.Fatal(err)
+	}
+
+	subdir3 := filepath.Join(toplevel, "part-3")
+	err = os.Mkdir(subdir3, 0750)
+	if err != nil && !os.IsExist(err) {
+		t.Fatal(err)
+	}
+
+	expected := []string{subdir1, subdir2, subdir3}
+	dirs, err := getConfigMapDirs(toplevel)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if diff := cmp.Diff(dirs, expected); diff != "" {
+		t.Fatal(diff)
+	}
+
+	// If we see a "..data" symlink, treat the toplevel file as the only
+	// ConfigMap dir. The target of the symlink must exist for the symlink to be
+	// created, so we just point to the toplevel folder itself; however the
+	// actual target itself doesn't matter for purposes of this test (we just
+	// check that getConfigMapDirs sees the "..data" entry and treats the
+	// toplevel folder as a single ConfigMap).
```

Comment:
Nit. Above, "The target of the symlink must exist for the symlink to be created," is not strictly true. Maybe you meant the symlink was created because a target directory was present? (Only commenting because I know you like to be precise. :-)

Reply:
> "The target of the symlink must exist for the symlink to be created," is not strictly true.

You are correct; my understanding of the root cause of why `os.Symlink("foo", data_dir)` (what I wanted to do originally) was making this test break was incorrect, and I've narrowed it down to our use of os.Stat() and not os.Lstat(). I now explain this in the comments.